### ISSUE TITLE
Feature request: Support static Razor compilation

### ISSUE THREAD

**Issue by jimasp** [OrchardCMS/Orchard#7713]
Is it possible to perform view compilation in Orchard?

Using the off-the-shelf latest dev branch, and setting `MvcBuildViews` to `true` results in all kinds of reference issues.

As an alternative, I tried doing this:

1. `build Precompiled` - this works fine, no problem
2. `%windir%\Microsoft.NET\Framework64\v4.0.30319\aspnet_compiler -f -v / -p c:\_src\orchard\build\Precompiled c:\temp\aspnetcompilerout`

But (2) fails, with the following error:

`c:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\fd596974\9555b8ff\App_Web_4psv3d0u.0.cs(34): error CS0246: The type or namespace name 'Lucene' could not be found (are you missing a using directive or an assembly reference?)`

Am I pointing the tool at the wrong folder? Is it possible to create a folder that would work from the targets?
Or is it just not possible to compile views in orchard without major divergence from the code base?

This guy seemed to manage it, and I think? i'm doing the same steps, although I don't have a "Web" folder, so his post is most likely out of date:
https://blogs.msdn.microsoft.com/shacorn/2014/11/13/asp-net-precompilation-speeding-up-orchard-cold-start/

(The reason i am doing this is not for performance, but because there has been the occasional view compilation problem, which has not shown up until well into the testing phase).

**Comment by sebastienros** [comment]
The web folder that is mentioned is the one where the application is deployed. It's up to you to name it `web` too or replace by the name you are using.

**Comment by jimasp** [comment]
yeah, i just misread that.

back to the question:
Is there any way to compile views? (without messing with lots of references). 

mvcbuildviews doesn't work due to reference issues.
aspnet_compiler doesn't work with the build\Precompiled output directory 

The aspnet_compiler failure seems really odd to me, as given that this is the folder that gets deployed, surely it must implicitly be compileable by the aspnet compiler on the web server, or the website wouldn't work?

**Comment by jimasp** [comment]
On further investigation, there are 2 problems:
A. a couple of minor view/ reference issues (**fixed in attached PR**)
B. the requirement that the aspnet compiler requires local copies of referenced dlls.

Publishing from Visual Studio works, but we cannot do this on a build machine.
Attempting to publish from the command line with the same publish properties file seems to fail:

e.g.
```
msbuild c:\_src\Orchard\src\Orchard.Web\Orchard.Web.csproj /p:PublishProfile=c:\_src\Orchard\src\Orchard.Web\Properties\PublishProfiles\publish1.pubxml /p:DeployOnBuild=true
```
results in:

```
"c:\_src\Orchard\src\Orchard.Web\Orchard.Web.csproj" (default target) (1) ->
"c:\_src\Orchard\src\Orchard.Web\Core\Orchard.Core.csproj" (default target) (5:2) ->
(AspNetPreCompile target) ->
  c:\_src\Orchard\src\Orchard.Web\Core\obj\Debug\AspnetCompileMerge\Source\web.config(33): error ASPCONFIG: Could not load file or assembly 'Orchard.Framework' or one of its dependencies. The system cannot find the file specified. [c:\_src\Orchard\src\Orchard
.Web\Core\Orchard.Core.csproj]
```

**Comment by jimasp** [comment]
If you skip the precompile in the publish:
`msbuild c:\_src\Orchard\src\Orchard.Web\Orchard.Web.csproj /p:PublishProfile=c:\_src\Orchard\src\Orchard.Web\Properties\PublishProfiles\publish1.pubxml /p:DeployOnBuild=true /p:PrecompileBeforePublish=false`

Then the publish succeeds,

But then if you run the aspnetcompiler against the folder created by the publish, e.g.
`%windir%\Microsoft.NET\Framework64\v4.0.30319\aspnet_compiler -f -v / -p c:\_src\Orchard\src\Orchard.Web\build\publish c:\_src\Orchard\src\Orchard.Web\build\publish\aspnetcompiled`

it fails.

Does anyone have an example of a working build script that aspnet_compiles the website that we can use on a build machine (i.e. without resorting to visual studio UI publish)?

**Comment by jimasp** [comment]
I managed to replicate the visual studio publish functionality with msbuild with the below command, (this can only be run after a successful build):

`msbuild src\orchard.web\orchard.web.csproj /p:PrecompileBeforePublish=true  /p:Configuration=Release /t:WebPublish /p:WebPublishMethod=FileSystem /p:publishUrl=..\..\build\publish /p:DeleteExistingFiles=true /p:DebugSymbols=true /p:EnableUpdateable=true`

But actually, having looked again at the files/folders created by either the command or visual studio publish, they only seem to precompile the top level directory (global.asax), not the modules. Also, if you change the EnableUpdatable flag to false, (either in visual studio publish wizard or the command), then the publish will fail.

Subsequently, I've come to believe that the simplest way to achieve view compilation is probably to change the orchard.proj to copy all the bin files around (copy all dlls in root bin folder to all mvc project bin folders - themes\bin, core\bin, modules\*\bin, and possibly vice versa) prior to calling an AspNetCompiler task (which may? need to be called for each mvc project folder), and then remove all the excess dlls created during the copy. I'm not sure it's worth all the effort though... may be better to just live with bugs in the views.

**Comment by BenedekFarkas** [comment]
I started to experiment with this today and just found this issue while looking for some tips to see if anyone bumped into the same errors as I did. Here's what I found on the latest `1.10.x` source:

For my testing, I'm running `MSBuild` on `Orchard.proj`, as that's the standard way to produce a deploy package, but this is just for the initial compilation:
```
msbuild /t:Compile Orchard.proj /p:Solution=src\Orchard.sln /m /tv:14.0 /p:VisualStudioVersion=14.0 /p:Configuration=Release /p:MvcBuildViews=true
```
I have `Visual Studio 2015`, so I also have `MSBuild 14.0`, but I found some warnings in the build logs, so I explicitly defined `MSBuildToolsVersion` (`tv` switch) and `VisualStudioVersion` to be sure that none of the additional files (like `target` files) used in the process fall back to an earlier version that may cause problems (those might be available on my system, since I had earlier versions of the tools installed).

First of all, `Orchard.Core.csproj` needs a slight change, because the project file and its parent folder does not have the same name, which causes errors when reaching the `AfterBuildCompiler` target. The original (`$(ProjectDir)\..\$(ProjectName)`) is evaluated to `Orchard.Core\Orchard.Core`:
```
<Target Name="AfterBuildCompiler" Condition="'$(MvcBuildViews)'=='true'">
    <AspNetCompiler VirtualPath="temp" PhysicalPath="$(ProjectDir)\..\$(ProjectName)" />
</Target>
```
Fixed:
```
<AspNetCompiler VirtualPath="temp" PhysicalPath="." />
```

After that, I have the following errors:

- ```E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\Orchard.Modules\web.config(33): error ASPCONFIG: Could not load file or assembly 'Orchard.Framework' or one of its dependencies. The system cannot find the file specified. [E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\Orchard.Modules\Orchard.Modules.csproj]```
- ```E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Core\web.config(33): error ASPCONFIG: Could not load file or assembly 'Orchard.Framework' or one of its dependencies. The system cannot find the file specified. [E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Core\Orchard.Core.csproj]```
- ```E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\TinyMce\web.config(33): error ASPCONFIG: Could not load file or assembly 'Orchard.Framework' or one of its dependencies. The system cannot find the file specified. [E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\TinyMce\TinyMce.csproj]```
- ```E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\Orchard.MultiTenancy\web.config(33): error ASPCONFIG: Could not load file or assembly 'Orchard.Framework' or one of its dependencies. The system cannot find the file specified. [E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\Orchard.MultiTenancy\Orchard.MultiTenancy.csproj]```
- ```d:\Users\Benedek\AppData\Local\Temp\Temporary ASP.NET Files\temp\3afb0f36\71066fcc\App_Web_myybko2i.0.cs(32): error CS0246: The type or namespace name 'Orchard' could not be found (are you missing a using directive or an assembly reference?) [E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\Lucene\Lucene.csproj]```
- ```d:\Users\Benedek\AppData\Local\Temp\Temporary ASP.NET Files\temp\3afb0f36\71066fcc\App_Web_myybko2i.0.cs(35): error CS0246: The type or namespace name 'Orchard' could not be found (are you missing a using directive or an assembly reference?) [E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Modules\Lucene\Lucene.csproj]```

The last two errors actually originate from `src\Orchard.Web\Modules\Lucene\Views\EditorTemplates\Parts.LuceneSettings.cshtml`.

@jimasp I think the reason you're not seeing more errors might be that you're building `Orchard.Web`, which only has a few project references (`Orchard.Framework` and `Orchard.Core` for example).

> the requirement that the aspnet compiler requires local copies of referenced dlls.

I'll try this on those projects that throw an error for me.

I'm still investigating and I'll comment here with my findings.

**Comment by BenedekFarkas** [comment]
The last two errors (related to `Lucene`) can be fixed by adding `<add assembly="Orchard.Framework" />` to its `web.config` (just like all the other modules). Then it throws the same error, as the other 4 modules above, but at least it's consistent. :)

**Comment by BenedekFarkas** [comment]
I managed to almost get a full compilation on the whole solution. All of the changes in @jimasp's PR are correct, plus there's one additional batch of change I had to do, which is that in every case, where there's a reference to a project in the `web.config` (which is needed for `Razor IntelliSense`), you also need to have the same project reference with `CopyLocal` set to `true`. I achieved by this by replacing `<Private>True</Private>` or `<Private>False</Private>` (or adding it if missing) for each project reference like that (usually `Orchard.Framework` and `Orchard.Core`; and `Orchard.Layouts` for a few projects) with `<Private>$(MvcBuildViews)</Private>` in every project.

After that all of the projects build successfully, except for `Orchard.Web`, which throws the following error:
```
/temp/Core/Orchard.Core.csproj(1,0): error ASPPARSE: Object reference not set to an instance of an object. [E:\DEV\Orchard\Orchard CMS\src\Orchard.Web\Orchard.Web.csproj]
```

Any ideas?

**Comment by jimasp** [comment]
The problem with setting MvcBuildViews=true and Private=true in the source is that it massively slows down the build in visual studio, and devs hate that.

An alternative would be to have the build script update the csproj files to update these attributes before running the compilation. (The small downside to that approach is that if the build script is run on a dev box, as opposed to a build machine, it changes all the project files - annoying if you have changes you were planning to commit). However, when I tried to do things via this method, I got the same ASPPARSE problem you did.

I then tried manually copying all the dll files from the root bin folder to a module bin folder and running the aspnet_compiler.exe directly against that module folder, and that worked just fine.

This is why I said:
"... the simplest way to achieve view compilation is probably to change the orchard.proj to copy all the bin files around (copy all dlls in root bin folder to all mvc project bin folders - themes\bin, core\bin, modules*\bin, and possibly vice versa) prior to calling an AspNetCompiler task (which may? need to be called for each mvc project folder), and then remove all the excess dlls created during the copy. I'm not sure it's worth all the effort though... may be better to just live with bugs in the views."

(Then I ran out of time to look at options and had to move on).

One alternative you could try would be to set Private=True in all projects (but leave MvcBuildViews=false) and then issue the publish command:

`msbuild src\orchard.web\orchard.web.csproj /p:PrecompileBeforePublish=true /p:Configuration=Release /t:WebPublish /p:WebPublishMethod=FileSystem /p:publishUrl=..\..\build\publish /p:DeleteExistingFiles=true /p:DebugSymbols=true /p:EnableUpdateable=false`

(I never tried that, it just occurred to me now).

**Comment by BenedekFarkas** [comment]
> The problem with setting MvcBuildViews=true and Private=true in the source is that it massively slows down the build in visual studio, and devs hate that.

I agree, but what I did is changing `<Private>` to have the same value as `MvcBuildViews`, so it only copies `Orchard.Framework.dll` and `Orchard.Core.dll` if you enable view compilation. Btw a lot of DLLs are set to `<Private>true</Private>` anyway (mostly the ones from NuGet packages I think).

> I then tried manually copying all the dll files from the root bin folder to a module bin folder and running the aspnet_compiler.exe directly against that module folder, and that worked just fine.

Actually, I think it would be pretty easy to add this to `Orchard.proj`. I still don't have any idea about the error on `Orchard.Web` (maybe @sebastienros can help? See my previous comment), but I think we can exclude that from this process anyway, as it doesn't views have on its own (and there shouldn't be any).

The publish command you mentioned looks interesting - I'll test it and see what's the difference between this and creating a deploy package through `Orchard.proj` (with and without view compilation).

**Comment by BenedekFarkas** [comment]
I've found a bunch of other things in multiple csprojs and web.configs, so I'll open a separate PR (it will include @jimasp's changes, so I can continue testing and improving my changes).

**Comment by BenedekFarkas** [comment]
I've submitted my PR (see above), but it's targeting `1.10.x` - I think it's safe to merge there, but some further changes might be required on `dev`.

Testing is quite easy: `.\ClickToBuild BuildViews` or `.\ClickToBuild BuildViews Orchard.proj src\CustomSolution.sln`.

**Comment by sebastienros** [comment]
Should it be part of the Precompiled target instead ?

**Comment by BenedekFarkas** [comment]
Could be, but: The `Precompiled` target calls 4 other targets: `Clean`, `Compiled`, `Package-Stage`, `Package-Precompiled`. `Compile` actually has calls to MSBuild, one of them has the `OutputPath` configured, which breaks view compilation for some reason (it can't find `System.Web.Mvc.dll` for a couple of projects, but I couldn't find out why), but the other one (which doesn't have `OutputPath` defined) works OK with view compilation. 
To be honest, I'm not really a fan of `Orchard.proj` for calling `MSBuild`, because it contains a lot of stuff that is only needed so that's easier to run on the public CI, while `MSBuild` should rather be called directly as it's more straightforward to configure than `ClickToBuild.cmd`. On the other hand, `Orchard.proj` is great for `Package-Stage` and `Package-Precompiled`.

Actually, if we change the `Compile` target to the following, then it's just a matter of passing the `MvcBuildViews` parameter to `MSBuild`:
```
  <Target Name="Compile">
    <!-- Compile to "OutputFolder" -->
    <MSBuild
      Projects="$(Solution)"
      Targets="Build"
      Properties="Configuration=$(Configuration);OutputPath=$(CompileFolder);MvcBuildViews=false" />
    <!-- Compile to "regular" output folder for devs using VS locally -->
    <MSBuild
      Projects="$(Solution)"
      Targets="Build"
      Properties="Configuration=$(Configuration);MvcBuildViews=$(MvcBuildViews)" />
  </Target>
```
Then you can call `MSBuild` like this:
```
msbuild /t:Compile Orchard.proj /m /p:Solution=src\Orchard.sln /p:MvcBuildViews=true
```
I'd still keep the `BuildViews` target in case you want to build your views for self-checking, but don't want to do anything else, but I admit it's a bit redundant.

**Comment by BenedekFarkas** [comment]
Modified codegen with the same changes as what was applied to existing project so Modules and Themes created with it automatically support view compilation as well. Changes are on 1.10.x - fully tested. If we can merge into dev, I'll continue updating modules there (there are a few more there).

**Comment by BenedekFarkas** [comment]
New PR for 1.10.x with additional changes: https://github.com/OrchardCMS/Orchard/pull/7892

**Comment by sebastienros** [comment]
Symlinks? In every project? I don't think it's sensible

**Comment by BenedekFarkas** [comment]
I admit that it's strange, but:

- They will only be generated if you use view compilation.
- With the current changes, they don't have any side effect on the build/deployment (you probably wouldn't use this output for a deployment package, but you can). Although now that I thought about it, the tools might need to be deployed, because JIT will probably need it. JIT might see a performance increase from at as well.
- There's not really other way, because the path, where the compiler executable can be found is baked in at ˙bin/roslyn/csc.exe˙ when `aspnet_compiler` is running (according to the documentation, there's no parameter to override that in the project, only system-wide for hosting). The only alternative is to copy the Roslyn tools to every folder (which happens for Orchard.Web, but for every project it's obviously an overkill).

Can you ask someone in the ASP.NET team (Taylor maybe)? I've been digging around SO and forums, but there's not really a use-case like ours.

**Comment by BenedekFarkas** [comment]
Static view compilation is (hopefully just) temporarily dropped in favor of supporting Roslyn (and thus C# 6 and above) for Dynamic Compilation. If and when https://github.com/aspnet/RoslynCodeDomProvider/issues/25 is fixed (PR: https://github.com/aspnet/RoslynCodeDomProvider/pull/26), we can set up the extension's web.configs with a simple change to be able build views statically.

**Comment by BenedekFarkas** [comment]
[Good news, everyone!](https://media.giphy.com/media/3o7abA4a0QCXtSxGN2/giphy.gif)

My PR (https://github.com/aspnet/RoslynCodeDomProvider/pull/26) was merged!
By accident. So it was reverted.

However, there was another set of changes added by the project authors to achieve the same thing in a different (and actually better) way. Already tested it and it's working great - requires minimal changes to Orchard (package version update and some web.config modification + the same stuff for codegen) to support.
I asked about when they're planning to release it - hopefully soon!

Test use-case: Dynamic Compilation and static View Compilation on Razor files that contain C# 7.2 language features.

**Comment by BenedekFarkas** [comment]
[New PR, wall of text, but it's all working with C# 7.3!](https://github.com/OrchardCMS/Orchard/pull/8103#issue-213059094)

**Comment by BenedekFarkas** [comment]
Merged into 1.10.x - dev will follow soon.
